proto/src/rr/rdata/tlsa.rs

```diff
@@ -343,17 +343,17 @@ impl TLSA {
 ///    /                                                               /
 ///    +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 /// ```
-pub fn read(decoder: &mut BinDecoder, rdata_length: u16) -> ProtoResult<TLSA> {
-    if rdata_length < 4 {
-        return Err("TLSA Resource Record too short".into());
-    }
-
-    let cert_usage = decoder.read_u8()?.into();
-    let selector = decoder.read_u8()?.into();
-    let matching = decoder.read_u8()?.into();
+pub fn read(decoder: &mut BinDecoder, rdata_length: Restrict<u16>) -> ProtoResult<TLSA> {
+    let cert_usage = decoder.read_u8()?.unverified().into();
+    let selector = decoder.read_u8()?.unverified().into();
+    let matching = decoder.read_u8()?.unverified().into();
 
     // the remaining data is for the cert
-    let cert_data = decoder.read_vec((rdata_length - 3) as usize)?;
+    let cert_len = rdata_length
+        .map(|u| u as usize)
+        .checked_sub(3)
+        .map_err(|_| ProtoError::from("invalid rdata length in TLSA"))?;
```

Comment:
@oherrala this uses the Restrict type to do the checked_sub and is inline with uses in other areas.